compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1397,12 +1394,12 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             let source_map = self.tcx.sess.source_map();
             let head_span = source_map.guess_head_span(span);
             if let Ok(head) = source_map.span_to_snippet(head_span) {
-                err.span_suggestion(head_span, "consider adding a derive", format!("#[derive(Default)]\n{head}"), Applicability::MaybeIncorrect);
+                err.subdiagnostic(ConsiderAddingADerive {
+                    span: head_span,
+                    suggestion: format!("#[derive(Default)]\n{head}")
+                });
```

Comment:
Why is this using `span_to_snippet`? can't this just be
```rust
ConsiderAddingADerive {
    span: head_span.shrink_to_lo(),
    suggestion: format!("#[derive(Default)]\n")
}
```

Reply:
Are you suggesting removing the surrounding `if` entirely?
There is no test for this currently and the else-case does seem unreachable given a sane span

Assuming that, I'm going to remove the second diag struct

Reply:
Yes, exactly.